File: cpp/kthLargest.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        return quickSelection(nums, 0, nums.size()-1, k-1);
    }
    
    int quickSelection(vector<int>& nums, int start, int end, int k) {
        if (start == end) return nums[start];
        int pivot = nums[end], storeIndex=start;
        
        for (int i=start; i < end; ++i) if (nums[i] > pivot) swap(nums, i, storeIndex++);
        swap(nums, end, storeIndex);
        
        if (storeIndex == k) {

            return nums[k];
        }
        if (storeIndex < k) return quickSelection(nums, storeIndex+1, end, k);
        return quickSelection(nums, start, storeIndex-1, k);
    }
    
    void swap(vector<int>& nums, int i, int storeIndex) {
        if (i==storeIndex) return;
        nums[i] ^= nums[storeIndex];
        nums[storeIndex] ^= nums[i];
        nums[i] ^= nums[storeIndex];
    }
};

int main(int argc, char const *argv[])
{
    Solution sol;
    vector<int> v;
    v.push_back(2);
    v.push_back(1);
    cout << sol.findKthLargest(v, 1);
    return 0;
}

File: kthLargest.py
class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        def quickSelect(start, end):
            pivot = nums[end]
            pivot_index = start
            
            for i in range(start, end):
                if nums[i] > pivot:
                    nums[pivot_index], nums[i] = nums[i], nums[pivot_index]
                    pivot_index += 1
            
            nums[pivot_index], nums[end] = nums[end], nums[pivot_index]
            
            if pivot_index == k - 1: return nums[pivot_index]
            
            if pivot_index < k - 1: return quickSelect(pivot_index+1, end)
            
            return quickSelect(start, pivot_index-1)
            
        return quickSelect(0, len(nums)-1)

            
sol = Solution()
print sol.findKthLargest([2,1], 1)